pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/OpenTelemetryTopicStats.java

```diff
@@ -0,0 +1,490 @@
+                .counterBuilder(COMPACTION_BYTES_COUNTER)
+                .setUnit("By")
+                .setDescription("The total size of the compacted entries.")
+                .buildObserver();
+
+        transactionCounter = meter
+                .upDownCounterBuilder(TRANSACTION_COUNTER)
+                .setUnit("{transaction}")
+                .setDescription("The number of transactions on this topic.")
+                .buildObserver();
+
+        delayedSubscriptionCounter = meter
+                .upDownCounterBuilder(DELAYED_SUBSCRIPTION_COUNTER)
+                .setUnit("{entry}")
+                .setDescription("The total number of message batches (entries) delayed for dispatching.")
+                .buildObserver();
+
+        batchCallback = meter.batchCallback(() -> pulsar.getBrokerService()
+                        .getTopics()
+                        .values()
+                        .stream()
+                        .map(topicFuture -> topicFuture.getNow(Optional.empty()))
+                        .forEach(topic -> topic.ifPresent(this::recordMetricsForTopic)),
+                subscriptionCounter,
+                producerCounter,
+                consumerCounter,
+                messageInCounter,
+                messageOutCounter,
+                bytesInCounter,
+                bytesOutCounter,
+                publishRateLimitHitCounter,
+                storageCounter,
+                storageLogicalCounter,
+                storageBacklogCounter,
+                storageOffloadedCounter,
+                backlogQuotaLimitSize,
+                backlogQuotaLimitTime,
+                backlogEvictionCounter,
+                backlogQuotaAge,
+                storageOutCounter,
+                storageInCounter,
+                compactionRemovedCounter,
+                compactionOperationCounter,
+                compactionDurationSeconds,
+                compactionBytesInCounter,
+                compactionBytesOutCounter,
+                compactionEntriesCounter,
+                compactionBytesCounter,
+                transactionCounter,
+                delayedSubscriptionCounter);
+    }
+
+    @Override
+    public void close() {
+        batchCallback.close();
+    }
+
+    private void recordMetricsForTopic(Topic topic) {
+        var topicName = TopicName.get(topic.getName());
+        var builder = Attributes.builder()
+                .put(OpenTelemetryAttributes.PULSAR_DOMAIN, topicName.getDomain().toString())
+                .put(OpenTelemetryAttributes.PULSAR_TENANT, topicName.getTenant())
+                .put(OpenTelemetryAttributes.PULSAR_NAMESPACE, topicName.getNamespace())
+                .put(OpenTelemetryAttributes.PULSAR_TOPIC, topicName.getPartitionedTopicName());
+        if (topicName.isPartitioned()) {
+            builder.put(OpenTelemetryAttributes.PULSAR_PARTITION_INDEX, topicName.getPartitionIndex());
+        }
+        var attributes = builder.build();
```

Comment:
@dragosvictor @merlimat Please note this contradicts all the work done to eliminate memory allocation. 
The Attributes for each topic should be persisted either in a Map or easier by saving it in the implementation of `Topic` interface you're getting. Then you simply retrieve it by `topic.getTopicOTelAttributes()`.
When you are not doing it, you are creating a memory allocation each time you run collect, for each topic. This is exactly what I was set out to avoid in the whole implementation rework of OpenTelemetry Java SDK memory mode.

Reply:
@asafm can you please create an issue to track this ?